### PR TITLE
2.4 Upgrade - Copy Mongo Space to Controller Config HA Space

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

The Mongo space name recorded against controller info will ultimately be superseded by the HA space name in controller config.

To ensure that an upgrade to Juju 2.4 does not break installations using Mongo space, we need to copy the old setting to the new location.

## QA steps

Accompanying unit tests.

## Documentation changes

None.

## Bug reference

Progresses https://bugs.launchpad.net/juju/+bug/1591962

### CODE REVIEW

File: state/upgrades.go [hunk 1]
```diff
@@ -1388,3 +1388,24 @@ func addRelationStatus(st *State) error {
...
+	if info.MongoSpaceState != MongoSpaceValid {
+		return nil
+	}
+
+	settings, err := readSettings(st.db(), controllersC, controllerSettingsGlobalKey)
+	if err != nil {
+		return errors.Annotate(err, "cannot get controller config")
+	}
+
+	settings.Set(controller.JujuHASpace, info.MongoSpaceName)
```
Comment (jameinel) [inline state/upgrades.go]: considering a desire to make things idempotent, should we check to see if something is already set?

what happens if both exist?

I think ultimately that controllerSettingsGlobalKey should win?


We might also want to check for "MongoSpaceValid" but the actual Space value is "".

File: state/upgrades_test.go [hunk 2]
```diff
@@ -2108,3 +2110,43 @@ func (s *upgradesSuite) TestAddRelationStatus(c *gc.C) {
...
+	err = s.state.setMongoSpaceState(MongoSpaceInvalid)
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = CopyMongoSpaceToHASpaceConfig(s.state)
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Check(getHASpaceConfig(s.state.db(), c), gc.Equals, "")
+}
+
+func getHASpaceConfig(db Database, c *gc.C) string {
```
Comment (jameinel) [inline state/upgrades_test.go]: Ultimately, isn't this something that we should have on State anyway?

File: state/upgrades.go (hunk 1 above)
Comment (manadart) [reply state/upgrades.go]: I did think of this; but assuming someone in the field is upgrading a controller to 2.4, the only way there could be a configured HA space would be if they bootstrapped with 2.4, as this is not a stable release. There should be no such config value... or have I got that wrong?

File: state/upgrades_test.go (hunk 2 above)
Comment (manadart) [reply state/upgrades_test.go]: Yes, there is actually no need to read it like this anyway. I will change it to use state.ControllerConfig(). As we only read it here.

File: state/upgrades.go [hunk 3]
```diff
@@ -1388,3 +1388,24 @@ func addRelationStatus(st *State) error {
 	})
 	return errors.Annotate(err, "adding relation status")
 }
+
+// CopyMongoSpaceToHASpaceConfig copies the Mongo space name from
+// ControllerInfo to the HA space name in ControllerConfig.
+func CopyMongoSpaceToHASpaceConfig(st *State) error {
```
Comment (ExternalReality) [inline state/upgrades.go]: I am starting to believe that you have a knack for writing exceptionally clear tests.

File: state/upgrades_test.go [hunk 4]
```diff
@@ -2108,3 +2110,43 @@ func (s *upgradesSuite) TestAddRelationStatus(c *gc.C) {
 		expectUpgradedData{statuses, expectedStatuses},
 	)
 }
+
+func (s *upgradesSuite) TestCopyMongoSpaceToHASpaceConfigWhenValid(c *gc.C) {
+	c.Assert(getHASpaceConfig(s.state.db(), c), gc.Equals, "")
```
Comment (ExternalReality) [inline state/upgrades_test.go]: I am starting to believe that you have a knack for writing exceptionally clear tests.

File: state/upgrades_test.go [hunk 5]
```diff
@@ -2108,3 +2110,39 @@ func (s *upgradesSuite) TestAddRelationStatus(c *gc.C) {
 		expectUpgradedData{statuses, expectedStatuses},
 	)
 }
+
+func (s *upgradesSuite) TestCopyMongoSpaceToHASpaceConfigWhenValid(c *gc.C) {
```
Comment (ExternalReality) [inline state/upgrades_test.go]: I am starting to believe that you have a knack for writing exceptionally clear tests.

**Comment by manadart** [comment]
Attempting another $$merge$$ - intermittent test failure pending on another PR.

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**